dev/WindowsAppRuntime_BootstrapDLL/MddBootstrap.cpp

```diff
@@ -337,10 +466,188 @@ CLSID FindDDLM(
         }
     }
     THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NO_MATCH), !foundAny, "AppExtension.Name=%ls, Major=%hu, Minor=%hu, Tag=%ls, MinVersion=%hu.%hu.%hu.%hu",
-                    appExtensionName, majorVersion, minorVersion, (!versionTag ? L"" : versionTag), minVersion.Major, minVersion.Minor, minVersion.Build, minVersion.Revision);
+                    appExtensionName, majorVersion, minorVersion, (!versionTag ? L"" : versionTag),
+                    minVersion.Major, minVersion.Minor, minVersion.Build, minVersion.Revision);
     return bestFitClsid;
 }
 
+void FindDDLMViaEnumeration(
+    UINT32 majorMinorVersion,
+    PCWSTR versionTag,
+    PACKAGE_VERSION minVersion,
+    std::wstring& ddlmPackageFamilyName,
+    std::wstring& ddlmPackageFullName)
+{
+    // Find the best fit
+    bool foundAny{};
+    PACKAGE_VERSION bestFitVersion{};
+    winrt::hstring bestFitPackageFamilyName{};
+    winrt::hstring bestFitPackageFullName{};
+
+    // We need to look for DDLM packages in the package family for release <major>.<minor> and <versiontag>
+    // But we have no single (simple) enumeration to match that so our logic's more involved compared
+    // to FindDDLMViaAppExtension():
+    // 1. Look for Framework packages with Name="microsoft.winappruntime.ddlm.<minorversion>.*[-shorttag]"
+    // 1a. Enumerate all Framework packages registered to the user
+    // 1b. Only consider packages whose Name starts with "microsoft.winappruntime.ddlm.<minorversion>."
+    // 1c. If versiontag is specified, Only consider packages whose Name ends with [-shorttag]
+    // 1d. Only consider packages whose PublisherID = "8wekyb3d8bbwe"
+    // 2. Check if the package is in the <majorversion>.<minorversion> release
+    // 2a. Check if the package's Description starts with "Microsoft Windows App Runtime DDLM <majorversion>.<minorversion> "
+    // 3. Check if the architecture matches
+    // 4. Check if the package meets the specified minVerrsion
+
+    const UINT16 majorVersion{ HIWORD(majorMinorVersion) };
+    const UINT16 minorVersion{ LOWORD(majorMinorVersion) };
+    WCHAR packageNamePrefix[PACKAGE_NAME_MAX_LENGTH + 1]{};
+    if (!g_test_ddlmPackageNamePrefix.empty())
+    {
+        FAIL_FAST_IF_FAILED(StringCchCopyW(packageNamePrefix, ARRAYSIZE(packageNamePrefix), g_test_ddlmPackageNamePrefix.c_str()));
+    }
+    else
+    {
+        wsprintf(packageNamePrefix, L"microsoft.winappruntime.ddlm.%hu.", minorVersion);
```

Comment:
consider wsprintf_s

Reply:
Q: Is wsprintf_s or StringCchPrintfW preferable these days?

There's a few references. I'll see about tweaking them in a future PR